database/database.go

```diff
@@ -1,127 +1,585 @@
+	localDatabase.wordsMap = nil
+	localDatabase.mux.UnlockWrite()
+	postgresDatabase.Close()
+}
+
+/* -------------------------------------------------------------------------- */
+/*                          EXPORTED QUOTES FUNCTIONS                         */
+/* -------------------------------------------------------------------------- */
+
+// GetQuotes returns a slice containing all quotes
+// The weight variable will be zero
+func GetQuotes() *[]QuoteT {
+	localDatabase.mux.LockRead()
+	defer localDatabase.mux.UnlockRead()
+	quoteSlice := localDatabase.quoteSlice
+
+	return &quoteSlice
+}
+
+// GetQuotesByString returns a slice containing all quotes
+// The weight variable will indicate how well the given text matches the corresponding quote
+func GetQuotesByString(text string) *[]QuoteT {
+
+	localDatabase.mux.LockRead()
+	defer localDatabase.mux.UnlockRead()
+
+	quoteSlice := localDatabase.quoteSlice
 
-	_, err := database.Exec("INSERT INTO teachers (name, title, note) VALUES ($1, $2, $3)", name, title, note)
+	for word, count := range getWordsFromString(text) {
+		wordsMapItem := localDatabase.wordsMap[word]
+
+		_ = count
+		for _, v := range wordsMapItem.occurenceSlice {
+			quoteSlice[v.enumid].Match += float32(v.count) / float32(wordsMapItem.totalOccurences)
+		}
+	}
+	return &quoteSlice
+}
+
+// StoreQuote stores a new quote
+// If the ID is not zero, StoreQuote will try to find the appropriate quote and overwrite it
+// If the ID is zero a new quote will be created
+func StoreQuote(q QuoteT) error {
+	var err error
+
+	// Verify connection to PostgreSQL database
+	err = postgresDatabase.Ping()
 	if err != nil {
-		log.Print("Cannot store teacher: ", err)
-		return err
+		postgresDatabase.Close()
+		return errors.New("At StoreQuote: " + err.Error())
+	}
+
+	if q.QuoteID == 0 {
+		// add quote to postgresDatabase
+		err = postgresDatabase.QueryRow(
+			`INSERT INTO quotes (TeacherID, Context, Text, Unixtime, Upvotes) VALUES ($1, $2, $3, $4, $5) RETURNING QuoteID`,
+			q.TeacherID, q.Context, q.Text, q.Unixtime, q.Upvotes).Scan(&q.QuoteID)
+		if err != nil {
+			return errors.New("At StoreQuote: " + err.Error())
+		}
+
+		// add quote to localDatabase
+		addQuoteToLocalDatabase(q)
+	} else {
+		// try to find corresponding entry postgresDatabase and overwrite it
+		var res sql.Result
+		res, err = postgresDatabase.Exec(
+			`UPDATE quotes SET TeacherID=$2, Context=$3, Text=$4, Unixtime=$5, Upvotes=$6 WHERE QuoteID=$1`,
+			q.QuoteID, q.TeacherID, q.Context, q.Text, q.Unixtime, q.Upvotes)
+		if err != nil {
+			return errors.New("At StoreQuote: " + err.Error())
+		}
+		if rowsAffected, _ := res.RowsAffected(); rowsAffected == 0 {
+			return errors.New("At StoreQuote: Could not find specified entry for overwrite")
+		}
+
+		// try to find corresponding entry in localDatabase and overwrite it
+		err = overwriteQuoteInLocalDatabase(q)
+		if err != nil {
+			return errors.New("At StoreQuote: " + err.Error())
```

Comment:
Failing to update the cache must not be a reason for this function to fail. Who knows why a teacher is not in the cache, perhaps the database was modified by another program.
Returning an error here would make the caller assume that the submission didn't make it to the database (though it did), causing them to prompt the user to re-enter the it, which we wouldn't want to happen. If updating the cache fails, just reload it - it must not be another burden on us.

Reply:
https://www.postgresql.org/docs/9.1/explicit-locking.html